R/touchstone_demographic_dataset.R

```diff
@@ -0,0 +1,245 @@
+                           e$db_tdd$dsource_code,
+                           e$db_tdd$dtype_code, sep = '\r')
+
+    # A mash to compare with existing demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$ds_mash <- paste(
+      e_tdd_csv$demographic_source,
+      e_tdd_csv$demographic_statistic_type,
+      sep = '\r')
+
+    e$db_dd$mash <- paste(e$db_dd$dsource_code,
+                          e$db_dd$dtype_code, sep = '\r')
+    e
+
+  } else {
+    list()
+  }
+}
+
+test_extract_touchstone_demographic_dataset <- function(e) {
+  if (!is.null(e$touchstone_demographic_dataset_csv)) {
+    e_tdd_csv <- e$touchstone_demographic_dataset_csv
+
+    testthat::expect_true(all(e_tdd_csv$demographic_source %in%
+                              e$db_dsrc$code),
+      label = "demographic sources in touchstone_demographic_dataset exist")
+
+    testthat::expect_true(all(e_tdd_csv$demographic_statistic_type %in%
+                    e$db_dstype$code),
+      label = "demog. statistic types in touchstone_demographic_dataset exist")
+
+    all_touchstones <- unique(c(e$tdd_touchstones$id, e$touchstone_csv$id))
+
+    testthat::expect_true(all(e_tdd_csv$touchstone %in% all_touchstones),
+      label = "Touchstones in touchstone_demographic_dataset exist")
+
+    testthat::expect_true(all(e_tdd_csv$ds_mash %in% e$db_dd$mash),
+      label = "Demographic datasets already exist")
+
+    testthat::expect_equal(0, sum(duplicated(e_tdd_csv$mash)),
+      label = "No duplicated (Touchstone, demographic_dataset) pair in csv")
+  }
+}
+
+###############################################################################
+
+transform_touchstone_demographic_dataset <- function(e) {
+
+  tdd <- e$touchstone_demographic_dataset_csv
+
+  if (!is.null(tdd)) {
+    tdd$demographic_dataset <- e$db_dd$id[match(tdd$ds_mash, e$db_dd$mash)]
+
+  # If any (touchstone,dataset) already exist in db's
+  # touchstone_demographic_dataset table, then they can be ignored.
+
+    tdd <- tdd[!tdd$mash %in% e$db_tdd$mash, ]
+
+  # If there is (touchstone,dataset) of same demographic_statistic_type
+  # then this should be an update (ie, to a new demog. source)
+  # and we need to look up demographic_dataset id for the updated
+  # source/type combo. (We assume this exists - part of demography
+  # update)
+
+    if (nrow(tdd) > 0) {
+      tdd$mash <- paste(tdd$touchstone, tdd$demographic_statistic_type,
+                      sep = '\r')
+
+      e$db_tdd$mash <- paste(e$db_tdd$touchstone,
+                             e$db_tdd$dtype_code, sep = '\r')
+
+      tdd$id <- e$db_tdd$id[match(tdd$mash, e$db_tdd$mash)]
+
+      which_nas <- which(is.na(tdd$id))
+      tdd$id[which_nas] <- seq(from = -1, by = -1,
+                               length.out = length(which_nas))
+
+      tdd$already_exists_db <- FALSE
+      tdd <- tdd[, c("already_exists_db", "demographic_dataset",
+                     "id", "touchstone")]
+      list(touchstone_demographic_dataset = tdd)
```

Comment:
you might find the early-exit/happy path optimising approach nicer to use here

Reply:
Done - yes, much nicer.

Reply:
Applied this a bit earlier in the extract too.